### PR TITLE
fix(auth): backport token blocklist update error handling

### LINKED ISSUE #2766: Blocklist update failures are silently swallowed on release-2.0 (missing backport of #2086)
### What is the issue?

`tokenStore.Add` in `pkg/session/token_store.go` declares `updateErr` for the `UpdateSecret` call but returns `err` from inside the `if updateErr != nil` branch. At that point `err` is guaranteed nil, because the earlier `GetSecret` already returned on failure — so a failed write to the blocklist Secret is reported as success.

The consequence is that server-side logout stops being reliable under exactly the class of transient error it was built to tolerate. `blocklist.Block` wraps `Add` in a `backoff.Retry` with 10 attempts, but a nil return makes the retry loop treat attempt 1 as successful, `POST /v1/auth/revoke` answers 204, and the access token is never blocklisted — it stays valid until it expires on its own.

Impact is bounded: the refresh token is revoked correctly through the separate `tokenRegistry.Revoke` path, so no new access tokens can be minted. The exposure window is the remaining `AccessTokenTTL` of the already-issued access token.

This is the same defect as #2085, fixed on `main` by #2086 and shipped in v1.16.0. It is present on `release-2.0` because that fix was never backported.

### Steps to Reproduce

Deterministically, as a unit test in `pkg/session`:

1. Build a `fakeclient` with an `interceptor.Funcs{Update: ...}` that returns a conflict error (`k8serrors.NewConflict`) for Secret writes.
2. Construct a `tokenStore` against it and call `Add(ctx, token)`.
3. Observe that `Add` returns nil, and that a following `Exists(ctx, token)` reports the token is not blocklisted.

At runtime:

1. Cause `UpdateSecret` on the blocklist Secret to fail transiently — a 409 from two concurrent logouts, or controller-runtime cache staleness.
2. Log out.
3. Replay the same access token against a protected endpoint; it is still accepted.

### Relevant Logs / Output

```go
secret = addDataToSecret(ts.l, secret, shortenedToken, time.Now().UTC())
_, updateErr := ts.client.UpdateSecret(ctx, secret)
if updateErr != nil {
	ts.l.Errorf("failed to update %s secret in the %s namespace with the %s shortened token, retrying: %v", secret.Name, secret.Namespace, shortenedToken, updateErr)
	return err // always nil here
}
return nil
```

The log line does fire, so the failure is visible in logs — it just never reaches the caller.

### OpenEverest Version

`release-2.0` (the v2 development branch) @ efd04062, and the `v2.0.0-dev.1` developer preview tag cut from it.

### Installation Method

Helm

### Kubernetes Distribution

Other (please specify below)

### Environment Details

Not environment-specific. The defect is in error propagation and reproduces on any distribution and any installation method.

No shipped release is affected. `main` and every v1.16.x tag carry the fix; the bug exists only on the v2 development line, where the sole tagged artifact is the `v2.0.0-dev.1` preview. Worth fixing now precisely so it never reaches a v2 release.

### Possible Solution

`return updateErr` instead of `return err`, matching what #2086 already did on `main`.

Note that `git cherry-pick 6727c51e` will not apply cleanly: the branches have diverged around this code (`release-2.0` has a 4-arg `newTokenStore` and `Add` reads `ts.namespace`, `main` has 3 args and hardcodes `common.SystemNamespace`), and the commit also carries a copyright-header hunk that is already present. The edit is easier applied by hand.

Two PRs are planned:

1. `release-2.0` — the fix plus a regression test. Going first, so the test is reviewed on the branch where it visibly goes red then green.
2. `main` — the same test, ported. The fix is already there, so this one closes the coverage gap only.

The test should assert the invariant rather than the return value: `Add` reports success only when the token is genuinely on the blocklist, verified by reading it back. A test that only asserts "an error came back" would still pass against an `Add` that returns the right error but writes nothing.

### Additional Context

`pkg/session/blocklist_test.go` has no case where `UpdateSecret` fails — the fake client's writes always succeed. That gap is why the original bug shipped, and why the missing backport went unnoticed: the suite on `release-2.0` is green.

Reported privately via the process in `SECURITY.md`. Handling it in the open rather than as a Security Advisory: the fix and its reasoning have been in a public commit since May, so an advisory protects nobody, and there is no v2 release behind this branch.

The broader concern is the backport gap itself — a fix that landed on `main` in May is still absent from the v2 development line nearly three months later, with no signal that it was missing.

Related: #2085 (original report), #2086 (fix on `main`).

### PR DESCRIPTION
**Pull request by Anuragkumar-687**
## What

Backports the fix from #2086 to `release-2.0`.

Part of #2766

`tokenStore.Add` incorrectly returned the `GetSecret` error instead of the `UpdateSecret` error, causing blocklist update failures to be reported as success and preventing the retry logic from running.

## Changes

- Return the `UpdateSecret` error from `tokenStore.Add`, matching the existing fix in #2086.
- Add a regression test covering `UpdateSecret` failures.

## Testing

Added `TestTokenStore_Add_UpdateFailure`, which verifies that:

- `Add` returns an error when the Secret update fails.
- A read-back using `Exists` confirms the token was not blocklisted.

Verified locally:

- `go test ./pkg/session/...`
- `go test -race ./pkg/session/...`

## Related

- Backports #2086

### CODE REVIEW

**Comment by vickysharma-prog** [comment]
@Anuragkumar-687 the backport against #2086 and against `release-2.0`.

The fix itself is right. The hunk in `token_store.go` is byte-identical to the one merged in #2086, which is what a backport should look like. #2766 suggests `return updateErr`; reassigning `err` to match the original commit is the better call for a backport, since the two branches now read the same.

The bug reproduces as described. When the update fails, `err` still holds the nil result of the earlier `GetSecret`, so `Add` returns nil on a failed write, and `blocklist.Block` wraps `Add` in `backoff.Retry`, so the nil ends the retry loop instead of driving it.

The test does the thing #2766 specifically asked for: it asserts the invariant by reading back through `Exists`, not just that an error came back. That distinction matters, since a variant returning the right error while writing nothing would pass the weaker check. It is also a real regression test on the unfixed code `Add` returns nil and `require.Error` fails.

The description says `Fixes #2766`, which will auto-close that issue on merge. #2766 plans two PRs: this one on `release-2.0`, and a second porting the same test to `main`, where the fix already exists but the coverage gap does not. Closing the issue here would lose the second one.

`Part of #2766` instead. Same ask @recharte made on #2617 for the same reason.

File: pkg/session/blocklist_test.go [hunk 1]
```diff
@@ -437,6 +439,63 @@ func TestBlocklist_IsBlocked(t *testing.T) {
...
+// TestTokenStore_Add_UpdateFailure asserts the invariant that Add reports
+// success only when the token is genuinely blocklisted. When the underlying
+// Secret update fails, Add must return the error rather than swallow it, and
+// the token must not be readable back via Exists.
+//
+// This is a regression test for the missing backport of #2086 (see #2766): on
+// the unfixed code Add returns the nil `err` from the earlier GetSecret instead
+// of the update error, so a failed write is reported as success and the token
+// is never blocklisted.
+func TestTokenStore_Add_UpdateFailure(t *testing.T) {
```
Comment (vickysharma-prog) [inline pkg/session/blocklist_test.go:451]: This targets `tokenStore.Add`, and the split in this package puts those in `token_store_test.go` (`TestCleanupOld`, `TestAddDataToSecret`) while `blocklist_test.go` holds the `TestBlocklist_*` cases. The name you've given it already says `TokenStore`, so it reads like it belongs in the other file. Non-blocking.

File: pkg/session/blocklist_test.go [hunk 2]
```diff
@@ -437,6 +439,63 @@ func TestBlocklist_IsBlocked(t *testing.T) {
...
+			},
+		})
+	k := kubernetes.NewEmpty(l, "test-ns").WithKubernetesClient(mockClient.Build())
+
+	store, err := newTokenStore(ctx, k, l, "test-ns")
+	require.NoError(t, err)
+
+	// The Secret write fails, so Add must surface the error and not report success.
+	err = store.Add(ctx, shortenedToken)
+	require.Error(t, err)
```
Comment (vickysharma-prog) [inline pkg/session/blocklist_test.go:492]: This pins that the error is returned, and the `Exists` read-back below covers the invariant, which is the part that matters. What is not covered is the consequence #2766 leads with, that the retry loop now actually runs. A case through `blocklist.Block` counting `tokenStore.Add` attempts on a persistent conflict would close that, though with `maxRetries = 10` at a 500ms constant backoff it would add about five seconds to the suite unless the interval is made injectable. Note rather than a request.

**Review by vickysharma-prog — CHANGES_REQUESTED** [review]
Two non-blocking notes inline. CI green, DCO signed.

**Comment by Anuragkumar-687** [comment]
Hi @recharte! The release-2.0 backport is ready for review whenever you have a chance. It includes the backported fix from #2086 along with the regression test we discussed in #2766. As you suggested, once this PR is merged, I'll open the follow-up PR against main to port the same regression test and close the coverage gap there. I'd really appreciate your feedback whenever you have time. Thanks!

File: pkg/session/blocklist_test.go [hunk 3]
```diff
@@ -437,6 +439,63 @@ func TestBlocklist_IsBlocked(t *testing.T) {
...
+
+	// Read-back invariant: because the write failed, the token must not be
+	// blocklisted. Asserting the return value alone would not catch a variant
+	// that reports success while writing nothing.
+	blocked, err := store.Exists(ctx, shortenedToken)
+	require.NoError(t, err)
+	assert.False(t, blocked)
```
Comment (recharte) [inline pkg/session/blocklist_test.go]: `Exists` reads `secret.Data`, and the fake client never populates it — it stores the `StringData` you hand it without doing the API server's StringData → Data conversion. So `blocked` is `false` here whether the write failed or succeeded, and this assertion holds unconditionally. Dropping the interceptor so `Add` fully succeeds: the persisted Secret comes back `Data=map[] StringData=map[list:9d1c1f98-…]` and `Exists` still reports `false`.

`TestBlocklist_Block` hit the same thing 240 lines up and says so — *"the mocked client does not do this StringData -> Data transformation in Secrets which the actual k8a API do, so we only check the StringData field"*. Same idiom here:

```suggestion
	// Read-back invariant: because the write failed, the token must not have
	// reached the Secret. As in TestBlocklist_Block, the fake client does not do
	// the StringData -> Data transformation the real API server does, so the
	// assertion has to be on StringData.
	secret, err := k.GetSecret(ctx, ctrlclient.ObjectKey{
		Name:      common.EverestBlocklistSecretName,
		Namespace: "test-ns",
	})
	require.NoError(t, err)
	assert.NotContains(t, secret.StringData[dataKey], shortenedToken)
```

File: pkg/session/blocklist_test.go (hunk 2 above)
Comment (recharte) [inline pkg/session/blocklist_test.go:492]: The defect was returning the wrong error variable, so "an error came back" is a notch weaker than the contract this is guarding — `return errors.New("boom")` satisfies it too. `kubernetes.UpdateSecret` returns the client error unwrapped, so the conflict can be pinned directly:

```suggestion
	err = store.Add(ctx, shortenedToken)
	require.Error(t, err)
	assert.True(t, k8serrors.IsConflict(err), "Add must surface the UpdateSecret error, got %v", err)
```

File: pkg/session/blocklist_test.go (hunk 1 above)
Comment (recharte) [reply pkg/session/blocklist_test.go:451]: Placement is fine as it is: `token_store_test.go` covers the pure helpers (`cleanupOld`, `addDataToSecret`) and never builds a client, whereas every fake-client-backed case and `mockNewBlocklist` live in this file. Moving it would drag the harness across for one test.

File: pkg/session/blocklist_test.go (hunk 2 above)
Comment (recharte) [reply pkg/session/blocklist_test.go:492]: `TestBlocklist_Block` already covers this: it deletes the Secret, calls `Block` again and asserts the `NotFound` surfaces, which only happens after `backoff.Retry` exhausts all 10 attempts. That case runs in 5.15s today. The only gap left is the same loop driven by an `UpdateSecret` failure rather than a `GetSecret` one, and that would cost a second 5s to pin a path the existing case already exercises. Not adding a test for it — no change needed here.

**Review by recharte — CHANGES_REQUESTED** [review]
One thing that isn't on a line: the description says `Fixes #2766`, so merging auto-closes it — and #2766 covers two PRs, this one and the port of the test to `main`. Could you change it to `Part of #2766`?

**Comment by Anuragkumar-687** [comment]
Thanks @recharte, addressed the review comments.

* Replaced the `Exists` read-back with a direct `Secret.StringData` check, since the fake client does not perform the `StringData -> Data` conversion.
* Strengthened the error assertion to verify that `Add` surfaces the `UpdateSecret` Conflict error, rather than just returning any error.
* Kept the test in `blocklist_test.go` and did not add another retry-loop test, as discussed.
* Updated the issue reference from `Fixes #2766` to `Part of #2766`.

`go test ./pkg/session/...` and `go test -race ./pkg/session/...` both pass.

**Review by recharte — APPROVED** [review]
Thanks @Anuragkumar-687 well done
